DRILL-7746: Add REST API Ability to Accept CSV Input

cgivre opened the pull request:
# [DRILL-7746](https://issues.apache.org/jira/browse/DRILL-7746): Add REST API Ability to Accept CSV Input

## Description
This test enables the REST plugin to query APIs that return CSV as well as JSON. 
## Documentation
This adds an additional configuration option for each connection: `inputType` which defaults to JSON. If the API outputs `csv` then user should set `inputType` to `csv`.  The plugin requires that the data be in CSV format with the first line being the column headers.  

This PR uses the same library as Drill currently does for CSV parsing, however it was not possible to use the `CompliantTextReader` without extensive modifications to that file.   This PR implements a simple CSV reader using the same underlying parser.

## Testing
Added unit test for CSV input.  Ran tests from CLI as well.

sanel commented on contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/HttpScanBatchCreator.java:
```diff
@@ -92,9 +92,16 @@ public void bind(ManagedScanFramework framework) { }
+          return new HttpCSVBatchReader(subScan);
+        } else {
+          return new HttpBatchReader(subScan);
+        }
```
Github won't let me leave the comment for `else` block down below, but that `return null;` can go as default (without `else`).

cgivre replied:
Fixed

sanel commented on contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/HttpCSVBatchReader.java:
```diff
@@ -0,0 +1,186 @@
+    // Case for empty result set
+    if (firstRow == null || firstRow.length == 0) {
+      return;
+    }
```
Can `columnWriters = new ArrayList<>();` go here (after `firstRow` checks)?

cgivre replied:
Sure can!  Fixed.

sanel commented on contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/HttpCSVBatchReader.java:
```diff
@@ -0,0 +1,186 @@
+    this.firstRow = csvReader.parseNext();
+
+    // Case for empty dataset
+    if (firstRow == null) {
```
Does it make sense to refactor complete function to this:

```java
this.firstRow = csvReader.parseNext();
if (firstRow != null) {
  for (String value : firstRow) {
      builder.addNullable(value, TypeProtos.MinorType.VARCHAR);
    }
}
return builder.buildSchema();
```

cgivre replied:
Indeed... Much cleaner.  Fixed.

sanel commented on contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/HttpCSVBatchReader.java:
```diff
@@ -0,0 +1,186 @@
+
+    // Build the Schema
+    builder = new SchemaBuilder();
+    TupleMetadata drillSchema = buildSchema();
```
If `buildSchema()` is called only once, I presume `csvReader.beginParsing(inStream)` can go in it, to localize state; something like:

```java
buildSchema(http.getInputStream());
```

cgivre replied:
I realized from this comment that I didn't implement the `close()` method.  While `buildSchema()` is only called once, the proper way to implement these readers is to explicitly close any inputstreams, readers or whatever.  I fixed that.

cgivre commented:
@sanel 
Thank you very much for the quick review!  I addressed your comments.  Please let me know if we're good to go.  If so I'll squash commits and rebase tomorrow.

sanel commented:
Great! Looks good.

cgivre commented:
Thanks @sanel! 
Commits squashed!
